app/src/main/java/com/example/plantsapp/data/firebase/repository/FirebasePlantsRepository.kt

```diff
@@ -0,0 +1,84 @@
+package com.example.plantsapp.data.firebase.repository
+
+import com.example.plantsapp.data.firebase.entity.FirebasePlant
+import com.example.plantsapp.domain.model.Plant
+import com.example.plantsapp.domain.repository.PlantsRepository
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.ktx.toObject
+import com.google.firebase.firestore.ktx.toObjects
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.tasks.await
+import timber.log.Timber
+import javax.inject.Inject
+
+class FirebasePlantsRepository @Inject constructor(
+    private val firestore: FirebaseFirestore
+) : PlantsRepository {
+
+    private val plantsCollection = firestore.collection(KEY_COLLECTION_PLANTS)
+
+    override fun observePlants(): Flow<List<Plant>> {
+        return callbackFlow {
+            val subscription = plantsCollection
+                .addSnapshotListener { value, error ->
+                    when {
+                        error != null -> {
+                            Timber.e("Listen failed..", error)
+                        }
+
+                        value != null -> {
+                            val plants = value
+                                .toObjects<FirebasePlant>()
+                                .map { it.toPlant() }
+                            trySend(plants)
+                        }
+                    }
+                }
+
+            awaitClose {
+                subscription.remove()
```

Comment:
So, this function is indeed called when the flow is closed or canceled(but I guess that is not the case here)
And calling cancel() is not necessary here

Reply:
Cool, thanks